org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/BundleUtilsTest.java

```diff
@@ -158,6 +158,171 @@ public void testLoadSameBundleMultipleTimes() throws Exception {
+
+		Bundle bundleA = context.getBundle(bundleALocation);
+		Bundle bundleB = context.getBundle(bundleBLocation);
+		Bundle bundleC = context.getBundle(bundleCLocation);
+
+		try {
+			assertTrue(bundleA.getState() == Bundle.STARTING || bundleA.getState() == Bundle.ACTIVE);
+			// non singleton bundle with same symbolic name and version should not be installed
+			assertNull(bundleB);
+			assertTrue(bundleC.getState() == Bundle.STARTING || bundleC.getState() == Bundle.ACTIVE);
+		} finally {
+			bundleA.uninstall();
+			bundleC.uninstall();
+		}
+	}
+
+	@Test
+	public void testLoadSameSingletonBundleFromDifferentLocation() throws Exception {
+		// First, we load a bundle depends on dependency bundle
+		String oldBundlePath = getBundle("testresources/extension-with-dependency-0.0.1", "jdt.ls.extension.with.dependency_0.0.1.jar");
+		String oldDependencyPath = getBundle("testresources/extension-with-dependency-0.0.1", "dependency_0.0.1.201911081535.jar");
+		BundleUtils.loadBundles(Arrays.asList(oldBundlePath, oldDependencyPath));
+		String oldBundleLocation = getBundleLocation(oldBundlePath, true);
+		String oldDependencyLocation = getBundleLocation(oldDependencyPath, true);
+
+		BundleContext context = JavaLanguageServerPlugin.getBundleContext();
+		Bundle oldBundle = context.getBundle(oldBundleLocation);
+		Bundle oldDependency = context.getBundle(oldDependencyLocation);
+
+		assertNotNull(oldBundle);
+
+		assertTrue(oldBundle.getState() == Bundle.STARTING || oldBundle.getState() == Bundle.ACTIVE);
+		assertTrue(oldDependency.getState() == Bundle.ACTIVE);
+		oldBundle.loadClass("jdt.ls.extension.with.dependency.Activator");
+		assertEquals(oldBundle.getState(), Bundle.ACTIVE);
+
+		// Now we load bundles from another location, this may happen when the LS extensions get updated.
+		String newBundlePath = getBundle("testresources/extension-with-dependency-0.0.2", "jdt.ls.extension.with.dependency_0.0.2.jar");
+		String newDependencyPath = getBundle("testresources/extension-with-dependency-0.0.2", "dependency_0.0.2.201911081538.jar");
+		BundleUtils.loadBundles(Arrays.asList(newBundlePath, newDependencyPath));
+
+		String newBundleLocation = getBundleLocation(newBundlePath, true);
+		String newDependencyLocation = getBundleLocation(newDependencyPath, true);
+
+		Bundle newBundle = context.getBundle(newBundleLocation);
+		Bundle newDependency = context.getBundle(newDependencyLocation);
+
+		try {
+			assertTrue(oldBundle.getState() == Bundle.UNINSTALLED);
+			assertTrue(oldDependency.getState() == Bundle.UNINSTALLED);
+
+			assertTrue(newBundle.getState() == Bundle.STARTING || oldBundle.getState() == Bundle.ACTIVE);
+			assertTrue(newDependency.getState() == Bundle.ACTIVE);
+			newBundle.loadClass("jdt.ls.extension.with.dependency.Activator");
+			assertEquals(newBundle.getState(), Bundle.ACTIVE);
+		} finally {
+			// Uninstall the bundle to clean up the testing bundle context.
+			newBundle.uninstall();

```

Comment:
BundleUtils.loadBundles(Arrays.asList(oldBundlePath, oldDependencyPath));
BundleUtils.loadBundles(Arrays.asList(newBundlePath, newDependencyPath));
probably throw uncaught exception, need extend the enclosing scope for try->finally.